[LOOP-1125] Upload pump data

darinkrauss opened the pull request:
- https://tidepool.atlassian.net/browse/LOOP-1125
- https://tidepool.atlassian.net/browse/LOOP-1124
- https://tidepool.atlassian.net/browse/LOOP-3389
- https://tidepool.atlassian.net/browse/LOOP-3388
- Allow scheduled basal rate on basal DoseEntry
- Add duringSuspend property to DoseEntry
- Persist duringSuspend property in CachedInsulinDeliveryObject
- Allow temporary basal during suspend to have scheduled basal rate
- Extend NewPumpEvent and DoseEntry to support RawRepresentable
- Update InsulinDeliveryStore to cache 90 days of date
- Add PumpAlarmType for alarm PumpEvents
- Add SyncPumpEvent to support RemoteDataService
- Combine dose and pump event data for RemoteDataService
- Update DoseStore.executePumpQuery to merge pump events and doses
- Add limit and inclusive to InsulinDeliveryStore.getDoseEntries
- Update model and model mapping
- Update MockPumpManager to allow non-dose pump events
- Update MockPumpManager settings to store pump alarms
- Add and update tests

Note: Needs review from @ps2.

Commit:
[LOOP-1125] Upload pump data

- https://tidepool.atlassian.net/browse/LOOP-1125
- https://tidepool.atlassian.net/browse/LOOP-1124
- https://tidepool.atlassian.net/browse/LOOP-3389
- https://tidepool.atlassian.net/browse/LOOP-3388
- Allow scheduled basal rate on basal DoseEntry
- Add duringSuspend property to DoseEntry
- Persist duringSuspend property in CachedInsulinDeliveryObject
- Allow temporary basal during suspend to have scheduled basal rate
- Extend NewPumpEvent and DoseEntry to support RawRepresentable
- Update InsulinDeliveryStore to cache 90 days of date
- Add PumpAlarmType for alarm PumpEvents
- Add SyncPumpEvent to support RemoteDataService
- Combine dose and pump event data for RemoteDataService
- Update DoseStore.executePumpQuery to merge pump events and doses
- Add limit and inclusive to InsulinDeliveryStore.getDoseEntries
- Update model and model mapping
- Update MockPumpManager to allow non-dose pump events
- Update MockPumpManager settings to store pump alarms
- Add and update tests

Commit:
Sync all pump events to insulin delivery store for simulated core data

Commit:
Make QueryAnchors conform to  Equatable

darinkrauss commented:
LW PR for build: https://github.com/tidepool-org/LoopWorkspace/pull/741

rickpasetto commented on LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift:
```diff
@@ -57,6 +60,9 @@ extension HKQuantitySample {
             if dose.type == .tempBasal {
                 metadata[MetadataKeyProgrammedTempBasalRate] = HKQuantity(unit: .internationalUnitsPerHour, doubleValue: dose.unitsPerHour)
             }
+            if dose.type == .suspend {
+                metadata[MetadataKeyDuringSuspend] = true
+            }
```
Should this be:

```suggestion
            metadata[MetadataKeyDuringSuspend] = dose.type == .suspend
```

if not, why not? Could `duringSuspend` below just be non-optional, like `manuallyEntered` is?  That would avoid awkward dealings with the optional bool.

darinkrauss replied:
No longer necessary. Removing `duringSuspend` per other PR feedback.

darinkrauss replied:
Actually, necessary here, but not in DoseEntry. Made non-optional per feedback.

rickpasetto commented on LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift:
```diff
@@ -110,6 +116,10 @@ extension HKQuantitySample {
         return metadata?[MetadataKeyProgrammedTempBasalRate] as? HKQuantity
     }
 
+    var duringSuspend: Bool? {
```
[nit] Consider making this non-optional (see above)

darinkrauss replied:
No longer necessary. Removing `duringSuspend` per other PR feedback.

darinkrauss replied:
Actually, necessary here, but not in DoseEntry. Made non-optional per feedback.

rickpasetto commented on LoopKit/InsulinKit/PumpEvent+CoreDataProperties.swift:
```diff
@@ -48,8 +49,10 @@ extension PumpEvent: Encodable {
         try container.encodeIfPresent(raw?.base64EncodedString(), forKey: .raw)
         try container.encodeIfPresent(title, forKey: .title)
+        try container.encodeIfPresent(insulinType, forKey: .insulinType)
+        try container.encodeIfPresent(automatic, forKey: .automatic)
```
interesting...`automatic` was missing!  Good catch.

rickpasetto commented on LoopKitTests/DoseStoreTests.swift:
```diff
@@ -593,194 +570,1042 @@ class DoseStoreQueryTests: PersistenceControllerTestCase {
+        wait(for: [completion], timeout: 10)
+    }
+
+    func testPagingOufOfOrderPumpEvents() {
```
[tiny nit] spelling
```suggestion
    func testPagingOutOfOrderPumpEvents() {
```

rickpasetto commented on LoopKit/InsulinKit/DoseStore.swift:
```diff
@@ -1646,49 +1627,121 @@ extension DoseStore {
+
+                var events: [SyncPumpEventWithModificationCounter] = []
+
+                // Walk the two sets of events and reconcile. As the events are sorted by start date, the event with the earlier
```
[nit] Consider splitting this logic out into separate function[s].  As it is, it is kinda hard to follow and get the big picture, all inside this one `executePumpQuery` function. The comments are helpful, though, thanks.

rickpasetto reviewed (approved):
Mostly nits.  It'd be really great to see the long `executePumpQuery` function refactored into smaller functions, but not blocking approval.

ps2 commented on LoopKit/InsulinKit/CachedInsulinDeliveryObject+CoreDataClass.swift:
```diff
@@ -111,6 +111,19 @@ class CachedInsulinDeliveryObject: NSManagedObject {
         }
     }
+
+    var duringSuspend: Bool? {
```
As discussed in Slack DMs...

darinkrauss replied:
Renamed to `isSuspend` and made non-optional. Removed from `DoseEntry`.

ps2 commented on LoopKit/InsulinKit/InsulinDeliveryStore.swift:
```diff
@@ -166,23 +166,24 @@ extension InsulinDeliveryStore {
     ///   - completion: A closure called once the dose entries have been retrieved.
     ///   - result: An array of dose entries, in chronological order by startDate, or error.
-    public func getDoseEntries(start: Date? = nil, end: Date? = nil, completion: @escaping (_ result: Result<[DoseEntry], Error>) -> Void) {
+    public func getDoseEntries(start: Date? = nil, end: Date? = nil, limit: Int? = nil, inclusive: Bool = true, completion: @escaping (_ result: Result<[DoseEntry], Error>) -> Void) {
```
Is this new inclusive flag used just for the test, or is it planned for some other client use?

darinkrauss replied:
It is required by the DoseStore when fetching doses for reconciling with pump events for upload to remote data services. It defaults to `true`, but in this case it must be set to `false`.

ps2 commented on LoopKit/InsulinKit/PumpAlarmType.swift:
```diff
@@ -0,0 +1,71 @@
+    case noPower
+    case occlusion
+    case other(_ details: String)
+    case overLimit
```
This (`overLimit`) seems pretty vague. Is there are particular kind of limit that this is intended for?

ps2 replied:
Maybe it's for expiration?  Having a pump expired alarm seems pretty useful. Would recommend changing the name in that case to be more explicit.

darinkrauss replied:
I'll remove it.

ps2 commented on LoopKit/InsulinKit/SyncPumpEvent.swift:
```diff
@@ -0,0 +1,27 @@
+
+import Foundation
+
+public struct SyncPumpEvent: Equatable {
```
`PersistedPumpEvent` is the type in LoopKit that is designed for representing pump events stored in DoseStore. Why is this being created and returned from queries instead?

darinkrauss replied:
Primarily because `PersistedPumpEvent` has a number of fields that don't need to be exposed to the remote data services. Plus, `type` is optional while it should really be required for remote data services. It isn't a big deal, but if you feel strongly I can replace it.

ps2 replied:
Exposing a protocol is a nice way to restrict visibility if you really need to hide internals from a public api. You can make PersistedPumpEvent As for `type`, what happens if you encounter a pump event where it is nil? What does that look like when being turned into a `SyncPumpEvent`? It does seem like the sole role of PersistedPumpEvent was exposing pump events to the query interface, so I think now we have two types that users of DoseStore will see that represent PumpEvents stored in DoseStore.

darinkrauss replied:
Specifically, what do you want me to do here.

darinkrauss replied:
Regarding your questions about when then pump event type is `nil`, it is explicitly ignored. The fact that the `type` property is optional appears to be an artifact of history. The `type` of a pump event will either be the explicitly specified type or the type of the explicitly specified dose. While it is technically possible from an API perspective to specify `nil` for both, it is not in actual usage (except in some tests). Therefore, it is an "invalid" pump event and, thus, ignored.

Commit:
Updates based upon PR feedback

- Rename optional duringSuspend to require with default isSuspend
- Remove duringSuspend from DoseEntry
- Refactor DoseStore.executePumpQuery
- Remove PumpAlarmType.overLimit
- Update tests

darinkrauss commented:
@ps2 @rickpasetto Ready for another review.

rickpasetto commented on LoopKit/InsulinKit/DoseStore.swift:
```diff
@@ -1640,86 +1640,8 @@ extension DoseStore {
-                    }
-                }
+                // Reconcile and limit the events
+                let events = self.limitPumpEvents(self.reconcilePumpEventsAndDoseEvents(pumpEvents: pumpEvents, doseEvents: doseEvents), limit: limit)
```
👍

rickpasetto commented on LoopKit/InsulinKit/DoseStore.swift:
```diff
@@ -1737,11 +1659,104 @@ extension DoseStore {
-        return dateFormatter
-    }()
+    // Walk the two sets of events and reconcile based upon start date and type.
+    private func reconcilePumpEventsAndDoseEvents(pumpEvents: [SyncPumpEventWithModificationCounter], doseEvents: [SyncPumpEventWithModificationCounter]) -> [SyncPumpEventWithModificationCounter] {
```
👍 And I like how you split the responsibilities with clear function params.

rickpasetto reviewed (approved):
LGTM (thanks for refactoring those functions)

ps2 reviewed (commented):
Just marking as reviewed, so I can get notified when the next iteration is ready that we discussed.

Commit:
Updates from PR feedback

- Restore separate dose and pump event uploads
- Upload doses from InsulinDeliveryStore
- Add modification counter to CachedInsulinDeliveryObject
- Remove SyncPumpEvent
- Remove PumpEvent and DoseEntry reconciliation

Commit:
Update DoseStore to handle mutable and deleted doses

- Move isMutable from PumpEvent to DoseEntry
- Mutable doses do not have deliveredUnits (as it is indeterminate)
- Purge mutable doses before adding pump events
- Do not allow manual doses to be mutable
- Move manual dose database access from DoseStore to InsulinDeliveryStore
- Mark deleted doses with tombstone and do not physically delete
- Update all DoseStore queries to properly handle mutable and deleted
- Allow mutable doses in InsulinDeliveryStore
- Do not persist mutable doses in HealthKit
- Update RemoteDataService.uploadDoseDate to include deleted doses
- Fix optional encoding of DoseEntry.automatic
- Update tests

darinkrauss commented:
@ps2 @rickpasetto Ready for another review. Hopefully this is the last one.

rickpasetto reviewed (commented):
These new changes look ok to me, but I defer to @ps2 for final say-so.

ps2 commented on LoopKit/InsulinKit/DoseStore.swift:
```diff
@@ -846,19 +836,19 @@ extension DoseStore {
-        guard doses.count > 0 else {
+    public func addDoses(_ doses: [DoseEntry], from device: HKDevice?, completion: @escaping (_ error: Error?) -> Void) {
+        guard doses.count > 0, !doses.contains(where: { $0.isMutable }) else {
             completion(nil)
```
Is it a programming error to call `addDoses` with a mutable dose? If so, should this be an assertion?

ps2 commented on LoopKit/InsulinKit/InsulinDeliveryStore.swift:
```diff
@@ -142,17 +156,22 @@ public class InsulinDeliveryStore: HealthKitSampleStore {
 
             self.handleUpdatedDoseData(updateSource: .queriedByHealthKit)
+
+            if dosesAdded {
```
Do we not want to notify if doses were deleted? If so, why not?

darinkrauss replied:
Sure, we can do that.

ps2 commented on LoopKit/InsulinKit/InsulinDeliveryStore.swift:
```diff
@@ -273,7 +335,7 @@ extension InsulinDeliveryStore {
     ///   - completion: A closure called once the dose entries have been stored.
     ///   - result: Success or error.
-    func addDoseEntries(_ entries: [DoseEntry], from device: HKDevice?, syncVersion: Int, manuallyEntered: Bool = false, completion: @escaping (_ result: Result<Void, Error>) -> Void) {
+    func addDoseEntries(_ entries: [DoseEntry], from device: HKDevice?, syncVersion: Int, resolveMutable: Bool = false, completion: @escaping (_ result: Result<Void, Error>) -> Void) {
```
Could you update the documentation to describe what use the `resolveMutable` parameter is for?

ps2 replied:
And is passing false (or using the default) necessary for some cases, or is it just an optimization for those times we don't need it?

darinkrauss replied:
This is just an optimization for when we don't need it. We don't want to resolve mutable dose entries *except* when we know the incoming dose entries will properly update any pre-existing mutable dose entries. This is only the case when called from `DoseStore.addPumpEvents`.

ps2 reviewed (approved):
LGTM. Mainly some questions. 

Also, does this still migrate from DIY master? I see the mapping was updated, so I'm guessing it does. And does it auto-migrate from the previous version?

Commit:
Updates based upon PR feedback

Commit:
Merge branch 'dev' into darinkrauss/LOOP-1125/upload-pump-data

# Conflicts:
#	LoopKit/InsulinKit/InsulinDeliveryStore.swift
#	LoopKit/Service/RemoteDataService.swift

Commit:
Fix issue with MockService
